File: day01/python/solve.py
import os
from collections import Counter


def get_file(filename):
    parent_directory = os.path.dirname(os.path.dirname(__file__))
    base_dir = os.path.dirname(parent_directory)

    return os.path.join(base_dir, "data", filename)


filename = "input.txt"
filepath = get_file(filename)

with open(filepath) as file:
    data = file.readlines()

    list1 = [int(ln.strip().split("   ")[0]) for ln in data]
    list2 = [int(ln.strip().split("   ")[1]) for ln in data]

# Part 1
list1.sort()
list2.sort()

diff = [abs(ln1 - ln2) for ln1, ln2 in zip(list1, list2)]

print(sum(diff))

# Part 2
counter = Counter(list2)
simm = [num * counter[num] for num in list1]

print(sum(simm))

File: day01/Go/solve.go
package main

import (
	"fmt"
	"math"
	"os"
	"sort"
	"strconv"
	"strings"
)

func main() {

	// filename := "teste.txt"
	filename := "../../data/input.txt"

	data, _ := os.ReadFile(filename)
	lines := strings.Split(strings.TrimSpace(string(data)), "\n")

	var left, right []int
	for _, line := range lines {
		numbers := strings.Fields(line)
		leftNum, _ := strconv.Atoi(numbers[0])
		rightNum, _ := strconv.Atoi(numbers[1])

		left = append(left, leftNum)
		right = append(right, rightNum)
	}

	// Part 1
	sort.Ints(left)
	sort.Ints(right)

	sum := 0
	for i := 0; i < len(left); i++ {
		sum += int(math.Abs(float64(left[i] - right[i])))
	}

	fmt.Println(sum)

	// Part 2
	counter := map[int]int{}

	for _, num := range right {
		counter[num] += 1
	}

	total := 0
	for _, num := range left {
		total += counter[num] * num
	}

	fmt.Println(total)
}
